exonum/src/messages2/authorization.rs

```diff
@@ -0,0 +1,145 @@
+use failure::Error;
+use hex::{FromHex, ToHex};
+
+use crypto::{
+    self, hash, Hash, PublicKey, SecretKey, Signature, PUBLIC_KEY_LENGTH, SIGNATURE_LENGTH,
+};
+
+use super::EMPTY_SIGNED_MESSAGE_SIZE;
+
+/// Correct raw message that was deserialized and verified, from `UncheckedBuffer`;
+/// inner data should be formed according to the following layout:
+/// | Position | Stored data |
+/// | - - - - - - - -| - - - - - - |
+/// | `0..32`  | author's PublicKey     |
+/// | `32`     | message class          |
+/// | `33`     | message type           |
+/// | `34..N`  | Payload                |
+/// | `N..N+64`| Signature                |
+///
+///
+/// Every creation of `SignedMessage` lead to signature verification, or data signing procedure,
+/// which can slowdown your code. Beware `SignedMessage` message, this procedure is not free.
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub struct SignedMessage {
+    pub(in messages) raw: Vec<u8>,
+}
+
+impl SignedMessage {
+    pub(crate) fn new(
+        cls: u8,
+        tag: u8,
+        value: Vec<u8>,
+        author: PublicKey,
+        secret_key: &SecretKey,
+    ) -> SignedMessage {
+        let mut buffer = Vec::new();
+        buffer.extend_from_slice(author.as_ref());
+        buffer.push(cls);
+        buffer.push(tag);
+        buffer.extend_from_slice(value.as_ref());
+        let signature = Self::sign(&buffer, secret_key).expect("Couldn't form signature");
+        buffer.extend_from_slice(signature.as_ref());
+        SignedMessage { raw: buffer }
+    }
+
+    /// Creates `SignedMessage` wrapper from the raw buffer.
+    /// Checks binary format and signature.
+    pub fn verify_buffer(buffer: Vec<u8>) -> Result<Self, Error> {
+        if buffer.len() <= EMPTY_SIGNED_MESSAGE_SIZE {
+            bail!("Message too short message_len = {}", buffer.len())
+        }
+        let signed = SignedMessage { raw: buffer };
+
+        let pk = signed.author();
+        let signature = signed.signature();
+
+        Self::verify(signed.data_without_signature(), &signature, &pk)?;
+
+        Ok(signed)
+    }
+
+    fn data_without_signature(&self) -> &[u8]{
+        debug_assert!(self.raw.len() > EMPTY_SIGNED_MESSAGE_SIZE);
+        let sign_idx = self.raw.len() - SIGNATURE_LENGTH;
+        &self.raw[0..sign_idx]
+    }
+
+    #[cfg(test)]
+    pub(crate) fn unchecked_from_vec(buffer: Vec<u8>) -> Self {
+        SignedMessage { raw: buffer }
+    }
+
+    #[cfg(not(test))]
+    pub(in messages) fn unchecked_from_vec(buffer: Vec<u8>) -> Self {
+        SignedMessage { raw: buffer }
+    }
```

Comment:
@ilammy Thank you for your argumentation. I remove `#[cfg(not(test))]` one.

>I'd also mark the unchecked_*() methods as unsafe to make their users think more closely whether they need them.

Calling this methods didn't lead to memory unsafety

Reply:
Strictly speaking, it will not lead to unsafe memory operations, but it may, for example, cause `author()` and other getters to panic if `SignedMessage` is constructed from an invalid buffer. `std::char::from_u32_unchecked()` illustrates the same principle: while passing it invalid arguments does not immediately lead to a crash or undefined behavior, doing so breaks the contract expected from `char`.

I do not insist on `unsafe`, `unchecked` is enough of a warning. Just adding my two cents and some historical precedents.